glib-macros/src/gboxed_derive.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use proc_macro2::TokenStream;
+use proc_macro_error::abort_call_site;
+use quote::quote;
+
+use crate::utils::parse_type_name;
+
+pub fn impl_gboxed(input: &syn::DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    let gtype_name = match parse_type_name(&input, "gboxed") {
+        Ok(v) => v,
+        Err(e) => abort_call_site!(
+            "{}: derive(GBoxed) requires #[gboxed(type_name = \"BoxedTypeName\")]",
+            e
+        ),
+    };
+
+    quote! {
+        impl BoxedType for #name {
+            const NAME: &'static str = #gtype_name;
+
+            fn get_type() -> glib::Type {
+                static mut TYPE_: glib::Type = glib::Type::Invalid;
+                static ONCE: ::std::sync::Once = ::std::sync::Once::new();
+
+                ONCE.call_once(|| {
+                    let type_ = glib::subclass::register_boxed_type::<Self>();
+                    unsafe {
+                        TYPE_ = type_;
+                    }
+                });
+
+                unsafe { TYPE_ }
+            }
+        }
+
+        impl glib::StaticType for #name {
+            fn static_type() -> glib::Type {
+                <#name as glib::subclass::boxed::BoxedType>::get_type()
+            }
+        }
+
+        impl glib::value::SetValue for #name {
+            unsafe fn set_value(value: &mut glib::value::Value, this: &Self) {
+                let ptr: *mut #name = Box::into_raw(Box::new(this.clone()));
+                glib::gobject_sys::g_value_take_boxed(
+                    glib::translate::ToGlibPtrMut::to_glib_none_mut(value).0,
+                    ptr as *mut _,
+                );
+            }
+        }
+
+        impl glib::value::SetValueOptional for #name {
+            unsafe fn set_value_optional(value: &mut glib::value::Value, this: Option<&Self>) {
+                let this = this.expect("None not allowed");
+                let ptr: *mut #name = Box::into_raw(Box::new(this.clone()));
+                glib::gobject_sys::g_value_take_boxed(
+                    glib::translate::ToGlibPtrMut::to_glib_none_mut(value).0,
+                    ptr as *mut _,
+                );
+            }
+        }
+
+        impl<'a> glib::value::FromValueOptional<'a> for &'a #name {
+            unsafe fn from_value_optional(value: &'a glib::value::Value) -> Option<Self> {
+                let ptr = glib::gobject_sys::g_value_get_boxed(
+                    glib::translate::ToGlibPtr::to_glib_none(value).0,
```

Comment:
This is problematic as it assumes that `glib` is available in scope with that specific name. That's not a given.

It would be better to use `$crate::glib` or so here (and if needed re-export `glib` from the crate root with `#[doc(hidden)]`). I don't know how that works with procedural macros though

Reply:
How is it different from the other `glib::` usages in the existing `GEnum` macro?

`glib-macro` is a dep of `glib` so I'm not sure how to re-export `glib` from the macro crate without introducing a circular dep.

(Btw, isn't this only a problem when not using edition 2018 as external crate are automatically in scope with 2018?)

Reply:
In 2018 edition you can rename crates in your `Cargo.toml`, which I usually do for GStreamer for example (`gstreamer` -> `gst`).

I don't know how that would work with procedural macros. Maybe you can take a look at `failure` / `failure_derive`, which should have the same problem (the derived trait is part of the `failure` crate).

Reply:
They just uses `::failure`, see https://github.com/rust-lang-nursery/failure/blob/master/failure_derive/src/lib.rs#L76

[structop-dervive does the same](https://github.com/TeXitoi/structopt/blob/master/structopt-derive/src/lib.rs#L84) so I'm afraid there is no magic solution here.

Reply:
Then let's do the same here I guess. `::glib`

Reply:
I.e. adding some leading `::` :)

Reply:
That breaks the doc example and I didn't find any workaround to make it work.
Any idea? I pushed the updated branch.

```
error[E0433]: failed to resolve: could not find `glib` in `{{root}}`
   --> src/subclass/boxed.rs:139:43
    |
139 |     #[derive(Clone, Debug, PartialEq, Eq, glib::GBoxed)]
    |                                           ^^^^^^^^^^^^ could not find `glib` in `{{root}}`
```

Reply:
That's a question for @GuillaumeGomez :) Doesn't the doc example have an `extern crate glib` in it and thus get `glib` at its root?

Reply:
No, tried that and got: `can't find crate`

Reply:
Intriguing. I'll try to take a look today (but don't hesitate to ping me so I don't forget :) ).

Reply:
@GuillaumeGomez ping? :)

Reply:
I'm completely overloaded... Will try to take a look this evening. Keeping the tab open.